### PR TITLE
Add MultiTypeMappingTransformation and test

### PR DESCRIPTION
**Pull request by AndreKurait**
### Description
Support Union resolution of MultiTypeMappingTransformation in metadata migrations.

Adds argument `--multi-type-behavior` allowing customers to specify UNION and migrate multi type indices.

* Category: New feature
* Why these changes are required? 6.8 multi type mapping support
* What is the old behavior before changes and new behavior after changes? Threw unsupported if index had multiple types

### Issues Resolved
[List any issues this PR will resolve]

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
Added E2E unit test

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: MetadataMigration/src/test/java/org/opensearch/migrations/MultiTypeMappingTransformationTest.java [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+            var snapshotCreator = new FileSystemSnapshotCreator(
+                    snapshotName,
+                    sourceClient,
+                    SearchClusterContainer.CLUSTER_SNAPSHOT_DIR,
+                    List.of(),
+                    snapshotContext.createSnapshotCreateContext()
+            );
+
+            // Get snapshot for ES 5
+            SnapshotRunner.runAndWaitForCompletion(snapshotCreator);
```
Comment (peternied) [inline MetadataMigration/src/test/java/org/opensearch/migrations/MultiTypeMappingTransformationTest.java:94]: Refactor all of the ES 5 & snapshot creation into a function to keep the flow cleaner.

File: MetadataMigration/src/test/java/org/opensearch/migrations/MultiTypeMappingTransformationTest.java [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+            // Restore snapshot
+            upgradedSourceOperations.restoreSnapshot(snapshotCreator.getRepoName(), snapshotCreator.getSnapshotName());
+
+            // Verify that the index exists on the upgraded cluster
+            var checkIndexUpgraded = upgradedSourceOperations.get("/" + originalIndexName);
+            assertThat(checkIndexUpgraded.getKey(), equalTo(200));
+            assertThat(checkIndexUpgraded.getValue(), containsString(originalIndexName));
+
+
+            upgradedSourceOperations.deleteAllSnapshotsAndRepository(snapshotCreator.getRepoName());
```
Comment (peternied) [inline MetadataMigration/src/test/java/org/opensearch/migrations/MultiTypeMappingTransformationTest.java:111]: Do we need this function? IMO is OK to have more than one snapshot repo in a cluster

File: MetadataMigration/src/test/java/org/opensearch/migrations/MultiTypeMappingTransformationTest.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+            // Use SnapshotImage as the transfer medium
```
Comment (peternied) [inline MetadataMigration/src/test/java/org/opensearch/migrations/MultiTypeMappingTransformationTest.java:113]: Nit: Http is uses less test code/resources and runs faster.

File: MetadataMigration/src/main/java/org/opensearch/migrations/MigrateOrEvaluateArgs.java [hunk 4]
```diff
@@ -89,4 +102,11 @@ public String getTransformerConfigParameterArgPrefix() {
                 description = "Path to the JSON configuration file of metadata transformers.")
         private String transformerConfigFile;
     }
+
+    static class MultiTypeResolutionBehaviorConverter implements IStringConverter<IndexMappingTypeRemoval.MultiTypeResolutionBehavior> {
```
Comment (peternied) [inline MetadataMigration/src/main/java/org/opensearch/migrations/MigrateOrEvaluateArgs.java:106]: Don't think you need this class, AFAIK enums are automatically handled

File: MetadataMigration/src/test/java/org/opensearch/migrations/MultiTypeMappingTransformationTest.java [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+            // Assert that both field1 and field2 are present
+            assertThat(properties.get("field1").get("type").asText(), equalTo("text"));
+            assertThat(properties.get("field2").get("type").asText(), equalTo("long"));
+            assertThat(properties.get("field3").get("type").asText(), equalTo("float"));
+        }
+    }
+
+    @SneakyThrows
+    @Test
+    public void multiTypeTransformationTest_union_withConflicts() {
```
Comment (peternied) [inline MetadataMigration/src/test/java/org/opensearch/migrations/MultiTypeMappingTransformationTest.java:180]: Don't create a full test for this edge case, write a unit test.

File: MetadataMigration/src/test/java/org/opensearch/migrations/MultiTypeMappingTransformationTest.java [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+            // Verify that the transformed index exists on the target cluster
+            var res = targetOperations.get("/" + originalIndexName);
+            assertThat(res.getKey(), equalTo(404));
+
+        }
+    }
+
+    @Data
+    @Builder
+    private static class TestMetadataTransformationParams implements MetadataTransformerParams {
```
Comment (peternied) [inline MetadataMigration/src/test/java/org/opensearch/migrations/MultiTypeMappingTransformationTest.java:297]: Unneeded, remove.  JCommand properties are always defined.

File: RFS/src/testFixtures/java/org/opensearch/migrations/bulkload/http/ClusterOperations.java [hunk 7]
```diff
@@ -122,7 +180,7 @@ public void takeSnapshot(final String snapshotName, final String indexPattern) t
             + indexPattern
             + "\",\n"
             + "  \"ignore_unavailable\": true,\n"
-            + "  \"include_global_state\": true\n"
+            + "  \"include_global_state\": false\n"
```
[the remark this answers is not in the document]
Comment (AndreKurait) [reply RFS/src/testFixtures/java/org/opensearch/migrations/bulkload/http/ClusterOperations.java]: Good catch, reverted

File: RFS/src/testFixtures/java/org/opensearch/migrations/bulkload/http/ClusterOperations.java [hunk 8]
```diff
@@ -53,19 +58,72 @@ public void createSnapshotRepository(final String repoPath) throws IOException {
...
+        // Delete the repository
+        HttpDelete deleteRepoRequest = new HttpDelete(clusterUrl + "/_snapshot/" + repoName);
+        deleteRepoRequest.setHeader("Content-Type", "application/json");
+        try (var response = httpClient.execute(deleteRepoRequest)) {
+            assertThat(response.getCode(), equalTo(200));
+        }
+    }
+
+    @SneakyThrows
+    public static void deleteSnapshotDir(SearchClusterContainer container) {
```
[the remark this answers is not in the document]
Comment (AndreKurait) [reply RFS/src/testFixtures/java/org/opensearch/migrations/bulkload/http/ClusterOperations.java:105]: Didn't need the logic anymore so deleted

File: RFS/src/testFixtures/java/org/opensearch/migrations/bulkload/http/ClusterOperations.java [hunk 9]
```diff
@@ -53,19 +58,72 @@ public void createSnapshotRepository(final String repoPath) throws IOException {
...
+        HttpGet getSnapshotsRequest = new HttpGet(clusterUrl + "/_snapshot/" + repoName + "/_all");
+        getSnapshotsRequest.setHeader("Content-Type", "application/json");
+
+        List<String> snapshotNames = new ArrayList<>();
+        try (var response = httpClient.execute(getSnapshotsRequest)) {
+            if (response.getCode() == 200) {
+                String responseBody = EntityUtils.toString(response.getEntity());
+                // Extract snapshot names from the response
+                int index = 0;
+                while ((index = responseBody.indexOf("\"snapshot\":\"", index)) != -1) {
```
[the remark this answers is not in the document]
Comment (AndreKurait) [reply RFS/src/testFixtures/java/org/opensearch/migrations/bulkload/http/ClusterOperations.java:72]: Didn't need the logic anymore so deleted

File: MetadataMigration/src/main/java/org/opensearch/migrations/MigrateOrEvaluateArgs.java (hunk 4 above)
Comment (AndreKurait) [reply MetadataMigration/src/main/java/org/opensearch/migrations/MigrateOrEvaluateArgs.java:106]: Removed, verified it's working without

File: MetadataMigration/src/test/java/org/opensearch/migrations/CustomTransformationTest.java [hunk 10]
```diff
@@ -239,43 +190,44 @@ private void performCustomTransformationTest(
...
         @Builder.Default
         private String transformerConfigParameterArgPrefix = "";
         private String transformerConfigEncoded;
         private String transformerConfig;
         private String transformerConfigFile;
     }
+
+    @Data
+    @Builder
+    private static class TestMetadataTransformationParams implements org.opensearch.migrations.bulkload.transformers.MetadataTransformerParams {
```
Comment (peternied) [inline MetadataMigration/src/test/java/org/opensearch/migrations/CustomTransformationTest.java:230]: Can you remove this?

File: transformation/src/test/java/org/opensearch/migrations/transformation/rules/IndexMappingTypeRemovalTest.java [hunk 11]
```diff
@@ -201,37 +228,102 @@ void testApplyTransformation_customTypes() {
         assertThat(indexJson.toPrettyString(), not(containsString(typeName)));
     }
 
-    @Test
-    void testApplyTransformation_twoCustomTypes() {
+    @ParameterizedTest
+    @CsvSource({
```
[the remark this answers is not in the document]
Comment (AndreKurait) [reply transformation/src/test/java/org/opensearch/migrations/transformation/rules/IndexMappingTypeRemovalTest.java:232]: I'm including None for `testApplyTransformation_twoCustomTypes` and `testApplyTransformation_twoMappingEntries`, let me know if there are any other tests i missed

File: RFS/src/testFixtures/java/org/opensearch/migrations/bulkload/http/ClusterOperations.java [hunk 12]
```diff
@@ -106,6 +124,17 @@ public void createIndex(final String index, final String body) {
         }
     }
 
+    @SneakyThrows
+    public String attemptCreateIndex(final String index, final String body) {
```
[the remark this answers is not in the document]
Comment (AndreKurait) [reply RFS/src/testFixtures/java/org/opensearch/migrations/bulkload/http/ClusterOperations.java]: Updated

**Review by peternied — APPROVED** [review]
